packages/expo/cli/start/server/BundlerDevServer.ts

```diff
@@ -0,0 +1,308 @@
+  ) {
+    const manifestType = options.forceManifestType || 'classic';
+    const Middleware = MIDDLEWARES[manifestType];
+    assert(Middleware, `Manifest middleware for type '${manifestType}' not found`);
+
+    const urlCreator = this.getUrlCreator();
+    const middleware = new Middleware(this.projectRoot, {
+      constructUrl: urlCreator.constructUrl.bind(urlCreator),
+      mode: options.mode,
+      minify: options.minify,
+      isNativeWebpack: this.name === 'webpack' && this.isTargetingNative(),
+    });
+    return middleware.getHandler();
+  }
+
+  /** Start the dev server using settings defined in the start command. */
+  public abstract startAsync(options: BundlerStartOptions): Promise<DevServerInstance>;
+
+  protected async postStartAsync(options: BundlerStartOptions) {
+    if (options.location.hostType === 'tunnel' && !APISettings.isOffline) {
+      await this._startTunnelAsync();
+    }
+    await this.startDevSessionAsync();
+
+    this.watchConfig();
+  }
+
+  protected abstract getConfigModuleIds(): string[];
+
+  protected watchConfig() {
+    const notifier = new FileNotifier(this.projectRoot, this.getConfigModuleIds());
+    notifier.startObserving();
+  }
+
+  /** Create ngrok instance and start the tunnel server. Exposed for testing. */
+  public async _startTunnelAsync(): Promise<AsyncNgrok | null> {
+    const port = this.getInstance()?.location?.port;
+    if (!port) return null;
+    Log.debug('[ngrok] connect to port: ' + port);
+    this.ngrok = new AsyncNgrok(this.projectRoot, port);
+    await this.ngrok.startAsync();
+    return this.ngrok;
+  }
+
+  protected async startDevSessionAsync() {
+    // This is used to make Expo Go open the project in either Expo Go, or the web browser.
+    // Must come after ngrok (`startTunnelAsync`) setup.
+
+    if (this.devSession) {
+      this.devSession.stop();
+    }
+
+    this.devSession = new DevelopmentSession(
+      this.projectRoot,
+      // This URL will be used on external devices so the computer IP won't be relevant.
+      this.isTargetingNative()
+        ? this.getNativeRuntimeUrl()
+        : this.getDevServerUrl({ hostType: 'localhost' })
+    );
+
+    await this.devSession.startAsync({
+      runtime: this.isTargetingNative() ? 'native' : 'web',
+    });
+  }
+
+  public isTargetingNative() {
+    // Temporary hack while we implement multi-bundler dev server proxy.
+    return true;
+  }
+
+  public isTargetingWeb() {
+    return false;
+  }
+
+  /**
+   * Sends a message over web sockets to any connected device,
+   * does nothing when the dev server is not running.
+   *
+   * @param method name of the command. In RN projects `reload`, and `devMenu` are available. In Expo Go, `sendDevCommand` is available.
+   * @param params
+   */
+  public broadcastMessage(
+    method: 'reload' | 'devMenu' | 'sendDevCommand',
+    params?: Record<string, any>
+  ) {
+    this.getInstance()?.messageSocket?.broadcast?.(method, params);
```

Comment:
Both messageSocket and its broadcast method are non-null on DevServerInstances:
```suggestion
    this.getInstance()?.messageSocket.broadcast(method, params);
```